src/inet/InetInterface.cpp

```diff
@@ -105,15 +109,41 @@ bool InterfaceIterator::Next()
     // TODO : Cleanup #17346
     return false;
 }
+
+InterfaceAddressIterator::InterfaceAddressIterator()
+{
+    mNetifAddrList = nullptr;
+    mCurAddr       = nullptr;
+}
+
 bool InterfaceAddressIterator::HasCurrent()
 {
-    return mIntfIter.HasCurrent();
+    return (mNetifAddrList != nullptr) ? (mCurAddr != nullptr) : Next();
 }
 
 bool InterfaceAddressIterator::Next()
 {
-    // TODO : Cleanup #17346
-    return false;
+    if (mNetifAddrList == nullptr)
+    {
+        mNetifAddrList = otIp6GetUnicastAddresses(Inet::globalOtInstance);
+
+        if (mNetifAddrList == nullptr)
+        {
+            return false;
+        }
+        mCurAddr = const_cast<otNetifAddress *>(mNetifAddrList);
+    }
+    else if (mCurAddr != nullptr)
+    {
+        mCurAddr = mCurAddr->mNext;
+    }
+
+    if (mCurAddr == nullptr)
+    {
+        return false;
+    }
+
+    return true;
```

Comment:
wondering if this shorter version is the same:

```cpp
if (mNetifAddrList == nullptr) {
  mNetifAddrList = otIp6GetUnicastAddresses(Inet::globalOtInstance);
}

if (mCurAddr == nullptr) {
  mCurAddr = mNetifAddrList;
} else {
  mCurAddr = mCurAddr->mNext;
}

return (mCurAddr != nullptr);
```

Reply:
Which makes me wonder if I need mNetifAddrlist ... it seems to only be used to initialize mCurAddr. Is the call to get the global list expensive? Otherwise:

```cpp
if (mCurAddr == nullptr) {
  mCurAddr = otIp6GetUnicastAddresses(Inet::globalOtInstance);
} else {
  mCurAddr = mCurAddr->mNext;
}

return (mCurAddr != nullptr);
```

Reply:
That would cause an infinite loop wouldn't it ? when we iterate through all available address, ` mCurAddr = mCurAddr->mNext` will set mCurAddr to `nullptr` and at the next call of `next()` we will restore the initial value with otIp6GetUnicastAddresses. With the current design the iterator wouldn't be able to "reset" itself at the cost of a single additional pointer.

Reply:
I would expect a `bool mInitialized` to be more intuitive than using a full pointer to mean the same.

Reply:
```cpp
if (!mInitialized) {
  mCurAddr = otIp6GetUnicastAddresses(Inet::globalOtInstance);
  mInitialized = true;
} else if (mCurAddr != nullptr) {
  mCurAddr = mCurAddr->mNext;
}

return (mCurAddr != nullptr);
```

Reply:
I think that matches with your current code, except the value of the mNetifAddrList is basically used only as a `is null` or `is not null` check.

Reply:
True, I'll leave it as is for now, just to get it merge before the weekend and open a follow up next week to swap for a bool.